pennylane_qulacs/qulacs_device.py

```diff
@@ -39,181 +74,239 @@
-
-            basis_state = 0
-            for bit in reversed(par[0]):
-                basis_state = (basis_state << 1) | bit
-
-            self._state.set_computational_basis(basis_state)
-        elif operation == 'QubitUnitary':
-            if len(par[0]) != 2 ** len(wires):
-                raise ValueError('Unitary matrix must be of shape (2**wires, 2**wires).')
-
-            unitary_gate = gate.DenseMatrix(wires, par[0])
-            self._circuit.add_gate(unitary_gate)
-        elif operation == 'Rot':
-            self._circuit.add_gate(gate.merge([
-                gate.RZ(wires[0], par[0]),
-                gate.RY(wires[0], par[1]),
-                gate.RZ(wires[0], par[2])
-            ]))
-        elif operation in ('CRZ', 'Toffoli', 'CSWAP'):
-            mapped_operation = self._operations_map[operation]
-            if callable(mapped_operation):
-                gate_matrix = mapped_operation(*par)
-            else:
-                gate_matrix = mapped_operation
+            self._state = QuantumState(self.num_wires)
 
-            dense_gate = gate.DenseMatrix(wires, gate_matrix)
-            self._circuit.add_gate(dense_gate)
-        else:
-            mapped_operation = self._operations_map[operation]
-            self._circuit.add_gate(mapped_operation(*wires, *par))
+        self._circuit = QuantumCircuit(self.num_wires)
 
-    @property
-    def state(self):
-        return self._state.get_vector()
+    def apply(self, operations, **kwargs):
+        rotations = kwargs.get("rotations", [])
 
-    def pre_measure(self):
-        self._circuit.update_quantum_state(self._state)
+        self.apply_operations(operations)
 
-    def expval(self, observable, wires, par):
-        bra = self._state.copy()
+        # Rotating the state for measurement in the computational basis
+        self.apply_operations(rotations)
 
-        if isinstance(observable, list):
-            A = self._get_tensor_operator_matrix(observable, par)
-            wires = [item for sublist in wires for item in sublist]
-        else:
-            A = self._get_operator_matrix(observable, par)
+    def apply_operations(self, operations):
+        """Apply the circuit operations to the state.
 
-        dense_gate = gate.DenseMatrix(wires, A)
-        dense_gate.update_quantum_state(self._state)
+        This method serves as an auxiliary method to :meth:`~.QulacsDevice.apply`.
 
-        expectation = inner_product(bra, self._state)
+        Args:
+            operations (List[pennylane.Operation]): operations to be applied
+        """
 
-        return expectation.real
+        for op in operations:
+            wires = op.wires
+            par = op.parameters
 
-    def probabilities(self):
-        states = itertools.product(range(2), repeat=self.num_wires)
-        probs = np.abs(self.state)**2
+            mapped_operation = self._operation_map[op.name]
+            if op.name.endswith(".inv"):
```

Comment:
might be better to use `if op.inverse:` here, rather than relying on the operation name